engine-spark/src/main/java/org/apache/kylin/engine/spark/SparkColumnCardinality.java

```diff
@@ -0,0 +1,158 @@
+            .isRequired(true).withDescription("Output").create(BatchConstants.ARG_OUTPUT);
+    public static final Option OPTION_PRJ = OptionBuilder.withArgName(BatchConstants.ARG_PROJECT).hasArg()
+            .isRequired(true).withDescription("Project name").create(BatchConstants.ARG_PROJECT);
+    public static final Option OPTION_COLUMN_COUNT = OptionBuilder.withArgName(BatchConstants.CFG_OUTPUT_COLUMN).hasArg()
+            .isRequired(true).withDescription("column count").create(BatchConstants.CFG_OUTPUT_COLUMN);
+
+    private Options options;
+
+    public SparkColumnCardinality() {
+        options = new Options();
+        options.addOption(OPTION_TABLE_NAME);
+        options.addOption(OPTION_OUTPUT);
+        options.addOption(OPTION_PRJ);
+        options.addOption(OPTION_COLUMN_COUNT);
+    }
+
+    @Override
+    protected Options getOptions() {
+        return options;
+    }
+
+    @Override
+    protected void execute(OptionsHelper optionsHelper) throws Exception {
+        String tableName = optionsHelper.getOptionValue(OPTION_TABLE_NAME);
+        String output = optionsHelper.getOptionValue(OPTION_OUTPUT);
+        int columnCnt = Integer.valueOf(optionsHelper.getOptionValue(OPTION_COLUMN_COUNT));
+
+        Class[] kryoClassArray = new Class[]{Class.forName("scala.reflect.ClassTag$$anon$1"),
+                Class.forName("org.apache.kylin.engine.mr.steps.SelfDefineSortableKey")};
+
+        SparkConf conf = new SparkConf().setAppName("Calculate table:" + tableName);
+        //set spark.sql.catalogImplementation=hive, If it is not set, SparkSession can't read hive metadata, and throw "org.apache.spark.sql.AnalysisException"
+        conf.set("spark.sql.catalogImplementation", "hive");
+        //serialization conf
+        conf.set("spark.serializer", "org.apache.spark.serializer.KryoSerializer");
+        conf.set("spark.kryo.registrator", "org.apache.kylin.engine.spark.KylinKryoRegistrator");
+        conf.set("spark.kryo.registrationRequired", "true").registerKryoClasses(kryoClassArray);
+
+        KylinSparkJobListener jobListener = new KylinSparkJobListener();
+        try (JavaSparkContext sc = new JavaSparkContext(conf)) {
+            sc.sc().addSparkListener(jobListener);
+            HadoopUtil.deletePath(sc.hadoopConfiguration(), new Path(output));
+            // table will be loaded by spark sql, so isSequenceFile set false
+            final JavaRDD<String[]> recordRDD = SparkUtil.hiveRecordInputRDD(false, sc, null, tableName);
+            JavaPairRDD<Integer, Long> resultRdd = recordRDD.mapPartitionsToPair(new BuildHllCounter())
+                    .reduceByKey((x, y) -> {
+                        x.merge(y);
+                        return x;
+                    })
+                    .mapToPair(record -> {
+                        return new Tuple2<>(record._1, record._2.getCountEstimate());
+                    })
+                    .sortByKey(true, 1);
+
+            if (resultRdd.count() == 0) {
```

Comment:
Both count and saveAsNewAPIHadoopFile are action of RDD, I think here should be cached to avoid recompute, am I right?